pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,414 @@
+type actuator struct {
+	gardenClient   kubernetes.Interface
+	replicaGetter  ReplicaGetter
+	replicaFactory ReplicaFactory
+	cfg            *config.ManagedSeedSetControllerConfiguration
+	recorder       record.EventRecorder
+	logger         *logrus.Logger
+}
+
+// NewActuator creates and returns a new Actuator with the given parameters.
+func NewActuator(
+	gardenClient kubernetes.Interface,
+	replicaGetter ReplicaGetter,
+	replicaFactory ReplicaFactory,
+	cfg *config.ManagedSeedSetControllerConfiguration,
+	recorder record.EventRecorder,
+	logger *logrus.Logger,
+) Actuator {
+	return &actuator{
+		gardenClient:   gardenClient,
+		replicaFactory: replicaFactory,
+		replicaGetter:  replicaGetter,
+		cfg:            cfg,
+		recorder:       recorder,
+		logger:         logger,
+	}
+}
+
+// Now returns the current local time. Exposed for testing.
+var Now = metav1.Now
+
+// Reconcile reconciles ManagedSeedSet creation or update.
+func (a *actuator) Reconcile(ctx context.Context, set *seedmanagementv1alpha1.ManagedSeedSet) (*seedmanagementv1alpha1.ManagedSeedSetStatus, error) {
+	// Initialize status
+	status := set.Status.DeepCopy()
+	status.ObservedGeneration = set.Generation
+
+	// Get replicas
+	replicas, err := a.replicaGetter.GetReplicas(ctx, set)
+	if err != nil {
+		return status, err
+	}
+
+	// Sort replicas by ascending ordinal
+	sort.Sort(ascendingOrdinal(replicas))
+
+	// Get the pending replica, if any
+	pendingReplica := getPendingReplica(replicas, status)
+
+	// Determine ready, postponed, and deletable replicas
+	var readyReplicas, postponedReplicas, deletableReplicas []Replica
+	var debug []string
+	for _, r := range replicas {
+		if replicaIsReady(r) {
+			readyReplicas = append(readyReplicas, r)
+		} else if r != pendingReplica {
+			postponedReplicas = append(postponedReplicas, r)
+		}
+		if r.IsDeletable() {
+			deletableReplicas = append(deletableReplicas, r)
+		}
+		debug = append(debug, replicaDebugString(r))
+	}
+	a.getLogger(set).Debugf("Replicas: %s", debug)
+
+	// Update replicas and readyReplicas in status
+	status.Replicas = int32(len(replicas))
+	status.ReadyReplicas = int32(len(readyReplicas))
+
+	// Determine the actual and target replica counts
+	count := len(replicas)
+	targetCount := 0
+	if set.DeletionTimestamp == nil {
+		targetCount = int(*set.Spec.Replicas)
+	}
+
+	// Determine whether scaling out or in
+	scalingOut, scalingIn := count < targetCount, count > targetCount
+
+	// Reconcile the pending replica, if any
+	if pendingReplica != nil {
+		if pending, err := a.reconcileReplica(ctx, set, status, pendingReplica, scalingIn); err != nil || pending {
+			return status, err
+		}
+	}
+
+	switch {
+	case scalingOut:
+		// Initialize a new replica and create its shoot
+		if err := a.createReplica(ctx, set, status, int(status.NextReplicaNumber)); err != nil {
+			return status, err
+		}
+
+		// Increment Replicas and NextReplicaNumber in status
+		status.Replicas++
+		status.NextReplicaNumber++
```

Comment:
If the status update at the end of the reconciliation fails, e.g. because of a conflict error or transient network issues, we cannot recover the next replica number, it simply gets lost.
I observed multiple cases, where this lead to the following error, from which the controller can't recover without a human manually patching the status:
```
time="2021-04-01T10:57:11+02:00" level=info msg="Error syncing ManagedSeedSet garden/gcp-eu: could not update status: ManagedSeedSet.seedmanagement.gardener.cloud \"gcp-eu\" is invalid: status.nextReplicaNumber: Invalid value: 0: cannot be less than status.replicas"
```

We should find some mechanism to detect which replica number we should use next even if we lost some status update.
One idea would be:
Before creating a new replica, check the highest ordinal of existing replicas and use at least one number higher than that if `NextReplicaNumber` lags behind.
This can still lead to reusing an ordinal in case the replica with the highest ordinal has been deleted.
Maybe you have a better idea?

Reply:
How did you manage to reproduce this, just by using cached runtime clients, or by also doing something else?

I don't have a better idea than what you propose. I will change the code so that `ordinal` is the greater of `NextReplicaNumber` and the highest existing replica ordinal + 1, and then set `NextReplicaNumber` to `ordinal+1` instead of simply incrementing it. I don't think reusing an ordinal is such a big issue, correct?

Reply:
Implemented as outlined above.